readme.md
# com.ugf.webrequests

UGF.WebRequests

## Information

- **Version**: `1.5.1`
- **Unity**: `2021.3`
- **API Compatibility Level**: `.NET Standard 2.1`

### Dependencies

- `com.ugf.initialize`: `2.9.0`
- `com.ugf.description`: `2.0.0`
- `com.ugf.logs`: `5.3.0`
- `com.ugf.runtimetools`: `2.15.0`
- `com.unity.modules.unitywebrequest`: `1.0.0`


### Description

Provides utilities to work with Web requests.

### Changelog

- [Full changelog](changelog.md)

### Registry

Read about where packages stored: [Registry](https://github.com/unity-game-framework/organization/blob/main/docs/registry.md).

## How to Install

Read about how to install any packages from the `Unity Game Framework` registry: [Install Package](https://github.com/unity-game-framework/organization/blob/main/docs/install-packages.md).


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.5.1](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.5.1) - 2022-08-15  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/11?closed=1)  
    

### Added

- Add request cookie check for value ([#25](https://github.com/unity-game-framework/ugf-webrequests/issues/25))  
    - Add `WebRequestCookie.HasValue` property to check whether value is specified or not.

## [1.5.0](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.5.0) - 2022-08-13  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/10?closed=1)  
    

### Added

- Add methods to work with cookies ([#23](https://github.com/unity-game-framework/ugf-webrequests/issues/23))  
    - Update dependencies: `com.ugf.runtimetools` to `2.15.0` version.
    - Add `IWebMessage.TryGetResponseCookies()`, `SetResponseCookies()`, `TryGetRequestCookies()` and `SetRequestCookies()` extension methods to work with cookies.
    - Add `WebRequestCookie` and `WebResponseCookie` structures to define, parse and format cookie data.
    - Add `WebRequestCookieCollection` and `WebResponseCookieCollection` classes as collection to work, parse and format combined cookie data.

## [1.4.0](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.4.0) - 2022-07-14  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/9?closed=1)  
    

### Added

- Add web request header dropdown ([#20](https://github.com/unity-game-framework/ugf-webrequests/issues/20))  
    - Update dependencies: `com.ugf.initialize` to `2.9.0`, `com.ugf.logs` to `5.3.0`, `com.ugf.runtimetools` to `2.9.2` and `com.ugf.editortools` to `2.8.1` versions.
    - Add `WebRequestHeaderDropdownAttribute` attribute used to display header selection text field with dropdown menu.

## [1.3.2](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.3.2) - 2022-03-22  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/8?closed=1)  
    

### Fixed

- Fix HttpWebRequestSender ([#18](https://github.com/unity-game-framework/ugf-webrequests/issues/18))  
    - Fix `HttpWebRequestSender` headers setup for request message.

## [1.3.1](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.3.1) - 2022-03-21  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/7?closed=1)  
    

### Fixed

- Fix HttpWebRequestSender content headers setup ([#15](https://github.com/unity-game-framework/ugf-webrequests/issues/15))  
    - FIx `HttpWebRequestSender` class request headers setup for body content.
- Fix HttpWebRequestSender max buffer size setup ([#14](https://github.com/unity-game-framework/ugf-webrequests/issues/14))  
    - Fix `HttpWebRequestSenderDescription.MaxResponseContentBufferSize` property to return `int` value.
    - Change `HttpWebRequestSenderDescription` class to have options to specific override for `Timeout` and `MaxBuffer` properties.

## [1.3.0](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.3.0) - 2022-03-18  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/6?closed=1)  
    

### Added

- Add unity web request send error message into response ([#12](https://github.com/unity-game-framework/ugf-webrequests/issues/12))  
    - Add error message in response as data when no data provided and error has occurred.

## [1.2.0](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.2.0) - 2022-03-02  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/5?closed=1)  
    

### Added

- Add sender and receiver builder assets ([#10](https://github.com/unity-game-framework/ugf-webrequests/issues/10))  
    - Update dependencies: `com.ugf.initialize` to `2.8.0` version.
    - Add `HttpWebRequestReceiverAsset` class as asset to build receiver for _Http_ solution.
    - Add `HttpWebRequestSenderAsset` class as asset to build sender for _Http_ solution.
    - Add `UnityWebRequestSenderAsset` class as asset to builder sender for _Unity_ solution.

## [1.1.3](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.1.3) - 2022-01-09  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/4?closed=1)  
    

### Fixed

- Fix to do not setup empty byte array as data ([#8](https://github.com/unity-game-framework/ugf-webrequests/issues/8))  
    - Change `HttpWebRequestReceiver`, `HttpWebRequestSender` and `UnityWebRequestSender` classes to skip data when it is an empty array.

## [1.1.2](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.1.2) - 2021-12-15  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/3?closed=1)  
    

### Fixed

- Fix UnityWebRequestSender set null data from download handler ([#7](https://github.com/unity-game-framework/ugf-webrequests/pull/7))  
    - Fix  `UnityWebRequestSender` class to check data from download handler when setup data for response.

## [1.1.1](https://github.com/unity-game-framework/ugf-webrequests/releases/tag/1.1.1) - 2021-11-29  

### Release Notes

- [Milestone](https://github.com/unity-game-framework/ugf-webrequests/milestone/2?closed=1)  
    

### Fixed

- Fix publish registry ([#5](https://github.com/unity-game-framework/ugf-webrequests/pull/5))  
    - Fix package publish registry.


Assets/UGF.WebRequests.Runtime.Tests/TestWebRequestCookieUtility.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using NUnit.Framework;

namespace UGF.WebRequests.Runtime.Tests
{
    public class TestWebRequestCookieUtility
    {
        private readonly List<(string value, WebResponseCookie cookie)> m_cookiesFormat = new List<(string value, WebResponseCookie)>
        {
            ("cookie=", new WebResponseCookie("cookie")),
            ("cookie=value", new WebResponseCookie("cookie", "value")),
            ("cookie=value; Expires=Wed, 21 Oct 2015 07:28:00 GMT; Max-Age=556460; Domain=test.com; Path=/path; Secure; HttpOnly; SameSite=Strict", new WebResponseCookie("cookie", "value")
            {
                Expires = DateTimeOffset.Parse("Wed, 21 Oct 2015 07:28:00 GMT"),
                MaxAge = TimeSpan.FromSeconds(556460D),
                Domain = "test.com",
                Path = "/path",
                Secure = true,
                HttpOnly = true,
                SameSite = WebCookieSameSite.Strict
            })
        };

        private readonly List<(string value, WebResponseCookie cookie)> m_cookiesValid = new List<(string value, WebResponseCookie)>
        {
            ("cookie", new WebResponseCookie("cookie")),
            ("cookie=", new WebResponseCookie("cookie")),
            ("cookie=value", new WebResponseCookie("cookie", "value")),
            ("cookie=value; Expires=Wed, 21 Oct 2015 07:28:00 GMT; Max-Age=556460; Domain=test.com; Path=/path; Secure; HttpOnly; SameSite=Strict", new WebResponseCookie("cookie", "value")
            {
                Expires = DateTimeOffset.Parse("Wed, 21 Oct 2015 07:28:00 GMT"),
                MaxAge = TimeSpan.FromSeconds(556460D),
                Domain = "test.com",
                Path = "/path",
                Secure = true,
                HttpOnly = true,
                SameSite = WebCookieSameSite.Strict
            }),
            ("cookie=value;Expires=Wed, 21 Oct 2015 07:28:00 GMT;Max-Age=556460;Domain=test.com;Path=/path;Secure;HttpOnly;SameSite=Strict", new WebResponseCookie("cookie", "value")
            {
                Expires = DateTimeOffset.Parse("Wed, 21 Oct 2015 07:28:00 GMT"),
                MaxAge = TimeSpan.FromSeconds(556460D),
                Domain = "test.com",
                Path = "/path",
                Secure = true,
                HttpOnly = true,
                SameSite = WebCookieSameSite.Strict
            }),
            ("cookie=value; Domain=test.com; Expires=Wed, 21 Oct 2015 07:28:00 GMT; HttpOnly; Max-Age=556460; SameSite=Strict; Path=/path; Secure", new WebResponseCookie("cookie", "value")
            {
                Expires = DateTimeOffset.Parse("Wed, 21 Oct 2015 07:28:00 GMT"),
                MaxAge = TimeSpan.FromSeconds(556460D),
                Domain = "test.com",
                Path = "/path",
                Secure = true,
                HttpOnly = true,
                SameSite = WebCookieSameSite.Strict
            }),
            ("cookie=value; expires=Wed, 21 Oct 2015 07:28:00 GMT; max-age=556460; domain=test.com; path=/path; secure; httponly; samesite=Strict", new WebResponseCookie("cookie", "value")
            {
                Expires = DateTimeOffset.Parse("Wed, 21 Oct 2015 07:28:00 GMT"),
                MaxAge = TimeSpan.FromSeconds(556460D),
                Domain = "test.com",
                Path = "/path",
                Secure = true,
                HttpOnly = true,
                SameSite = WebCookieSameSite.Strict
            }),
            ("cookie=value; expires=Wed, 21 Oct 2015 07:28:00 GMT; max-age=556460; domain=test.com; attribute=value; path=/path; secure; httponly; samesite=Strict", new WebResponseCookie("cookie", "value")
            {
                Expires = DateTimeOffset.Parse("Wed, 21 Oct 2015 07:28:00 GMT"),
                MaxAge = TimeSpan.FromSeconds(556460D),
                Domain = "test.com",
                Path = "/path",
                Secure = true,
                HttpOnly = true,
                SameSite = WebCookieSameSite.Strict
            })
        };

        [Test]
        public void ParseCookieValid()
        {
            foreach ((string value, WebResponseCookie cookie) in m_cookiesValid)
            {
                WebResponseCookie cookieResult = WebRequestCookieUtility.ParseCookie(value);

                AssertEqualCookie(cookie, cookieResult);
            }
        }

        [Test]
        public void ParseCookieInvalid()
        {
            Assert.Throws<ArgumentException>(() => WebRequestCookieUtility.ParseCookie(""));
            Assert.Throws<ArgumentException>(() => WebRequestCookieUtility.ParseCookie(";"));
            Assert.Throws<ArgumentException>(() => WebRequestCookieUtility.ParseCookie(";=value"));
            Assert.Throws<ArgumentException>(() => WebRequestCookieUtility.ParseCookie("cookie@=value"));
        }

        [Test]
        public void TryParseCookie()
        {
            bool result1 = WebRequestCookieUtility.TryParseCookie("sessionId=38afes7a8", out WebResponseCookie cookie1);
            bool result2 = WebRequestCookieUtility.TryParseCookie("id=a3fWa; Expires=Wed, 21 Oct 2015 07:28:00 GMT", out WebResponseCookie cookie2);
            bool result3 = WebRequestCookieUtility.TryParseCookie("id=a3fWa; Max-Age=2592000", out WebResponseCookie cookie3);
            bool result4 = WebRequestCookieUtility.TryParseCookie("qwerty=219ffwef9w0f; Domain=somecompany.co.uk", out WebResponseCookie cookie4);
            bool result5 = WebRequestCookieUtility.TryParseCookie("cook=value; Secure", out WebResponseCookie cookie5);

            Assert.True(result1);
            Assert.True(result2);
            Assert.True(result3);
            Assert.True(result4);
            Assert.True(result5);
            Assert.AreEqual("sessionId", cookie1.Name);
            Assert.AreEqual("38afes7a8", cookie1.Value);
            Assert.AreEqual("id", cookie2.Name);
            Assert.AreEqual("a3fWa", cookie2.Value);
            Assert.AreEqual("Wed, 21 Oct 2015 07:28:00 GMT", cookie2.Expires.ToString("R"));
            Assert.AreEqual("id", cookie3.Name);
            Assert.AreEqual(2592000D, cookie3.MaxAge.TotalSeconds);
            Assert.AreEqual("qwerty", cookie4.Name);
            Assert.AreEqual("219ffwef9w0f", cookie4.Value);
            Assert.AreEqual("somecompany.co.uk", cookie4.Domain);
            Assert.AreEqual("cook", cookie5.Name);
            Assert.AreEqual("value", cookie5.Value);
            Assert.True(cookie5.Secure);
        }

        [Test]
        public void FormatCookie()
        {
            foreach ((string value, WebResponseCookie cookie) in m_cookiesFormat)
            {
                string result = WebRequestCookieUtility.FormatCookie(cookie);

                Assert.AreEqual(result, value);
            }
        }

        [Test]
        public void ParseCookiePairs()
        {
            var collection = new List<(string Name, string Value)>
            {
                ("cookie", "value"),
                ("cookie", "value"),
                ("empty", "")
            };

            List<(string Name, string Value)> result = WebRequestCookieUtility.ParseCookiePairs("cookie=value; cookie=value; empty=");

            Assert.AreEqual(collection.Count, result.Count);

            for (int i = 0; i < result.Count; i++)
            {
                (string name, string value) = result[i];

                Assert.AreEqual(collection[i].Name, name);
                Assert.AreEqual(collection[i].Value, value);
            }
        }

        [Test]
        public void FormatCookiePairs()
        {
            var collection = new List<(string Name, string Value)>
            {
                ("cookie", "value"),
                ("cookie", "value"),
                ("empty", "")
            };

            string result = WebRequestCookieUtility.FormatCookiePairs(collection);

            Assert.AreEqual("cookie=value; cookie=value; empty=", result);
        }

        [Test]
        public void ParseCookieCollection()
        {
            string value = string.Join(',', m_cookiesValid.Select(x => x.value));

            List<WebResponseCookie> result = WebRequestCookieUtility.ParseCookieCollection(value);

            for (int i = 0; i < result.Count; i++)
            {
                AssertEqualCookie(m_cookiesValid[i].cookie, result[i]);
            }

            Assert.Pass($"Parsing cookie collection: {value}");
        }

        [Test]
        public void FormatCookieCollection()
        {
            string value = WebRequestCookieUtility.FormatCookieCollection(m_cookiesValid.Select(x => x.cookie).ToList());

            List<WebResponseCookie> result = WebRequestCookieUtility.ParseCookieCollection(value);

            for (int i = 0; i < result.Count; i++)
            {
                AssertEqualCookie(m_cookiesValid[i].cookie, result[i]);
            }

            Assert.Pass($"Parsing cookie collection: {value}");
        }

        private void AssertEqualCookie(WebResponseCookie first, WebResponseCookie second)
        {
            PropertyInfo[] properties = typeof(WebResponseCookie).GetProperties();

            foreach (PropertyInfo property in properties)
            {
                Assert.AreEqual(property.GetValue(first), property.GetValue(second), "{0}", property.Name);
            }
        }
    }
}
